### ISSUE TITLE
[feature] Add a `trim()` function to CORE

### ISSUE THREAD

**Issue by scottchiefbaker** [Perl/perl5#17952]
Perl is one of the strongest string manipulation languages around. Somehow Perl lacks a `trim()` function in CORE to remove white space. All modern string based languages have something similar: [PHP](https://www.php.net/manual/en/function.trim.php), [Javascript](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/trim), [Ruby](https://ruby-doc.org/core-2.7.1/String.html#method-i-strip), [Raku](https://docs.raku.org/routine/trim), and [Python](https://www.journaldev.com/23625/python-trim-string-rstrip-lstrip-strip). Can Perl please have one too? 

I propose adding `trim()` to CORE for the next release of Perl.

Relevant [Reddit converstion](https://www.reddit.com/r/perl/comments/hf3jlx/announcing_perl_7/fvwos56/) on the topic.

**Comment by scottchiefbaker** [comment]
It's worth noting that this feature **could** be implemented with a regular expression pretty easily. I feel strongly that the syntactical sugar of a `trim()` function would increase code readability tremendously.

**Comment by Grinnz** [comment]
Rough draft for docs for the function in perlfunc, based on discussions in #p5p regarding unicode semantics. It will need to be enabled by a feature guard.

```
=head2 trim EXPR

Returns a trimmed substring of EXPR with leading and trailing whitespace
removed. Equivalent to running the following, but more efficient:

  my $trimmed = ($str =~ s/\A\s+//ru) =~ s/\s+\z//ru;

Unicode semantics will be used to determine the whitespace to remove, which
means unicode space characters like U+00A0 "NO-BREAK SPACE" and U+2002 "EN
SPACE" will be trimmed as well.
```

**Comment by leonerd** [comment]
Idle thought: Would `trim` therefore be a regular "value in, value out"-style function, rather than an inplace mutator like `chomp`?

**Comment by scottchiefbaker** [comment]
@leonerd yes it needs to be "value in, value out" 100%. I find inplace mutators like `chomp()` to be less useful..

**Comment by Grinnz** [comment]
Also note my "but more efficient" is a bit of wishful thinking that should probably get verified once an actual implementation exists ;)

**Comment by scottchiefbaker** [comment]
PHP has an [interesting implementation](https://www.php.net/trim) of `trim()` that lets you choose the characters to remove, but also has sane defaults. I think maybe we should copy that implementation, and keep it simple.

**Comment by leonerd** [comment]
@Grinnz That seems a reasonable first draft at docs. Between you and @scottchiefbaker if you arrange for a branch containing that doc, and some tests (e.g. see most of the ones in `t/op/*` for starters), I can have a go at providing an actual impl. It'll mostly be just cargoculting my earlier `isa` change for the most part anyway.

**Comment by Grinnz** [comment]
A second optional argument to specify the characters to remove could be sensible. The perlish way that comes to mind is allowing a `qr//` argument, but allowing an arbitrary regex might be counterproductive when we really want just a character class.

**Comment by scottchiefbaker** [comment]
With all due respect, I think we need to rethink some of our documentation. A **lot** of the Perl documentation is very engineer heavy, and not very novice friendly. I think @Grinnz's sample above is a great example. The majority of people that are going to be looking at the documentation for a simple thing like `trim()` are going to be novices. They *probably* don't care about that sample regexp because it's over their head, and the unicode semantics are a pretty corner-case for this. 

I propose something simpler and more straight forward (borrowed heavily from the PHP trim) documentation:

## trim

`trim` EXPR  
`trim` EXPR,EXPR

Strip whitespace (or other characters) from the beginning and end of a string.

This function returns a string with whitespace stripped from the beginning and end of string. Without the second parameter, `trim()` will remove these characters: 

" ",  an ordinary space.  
"\t", a tab.  
"\n", a new line (line feed).  
"\r", a carriage return.  

#### Examples:

     $str = trim($str)
     $str = trim($str, ["\t", "\n", " "]);

**Comment by smpeters** [comment]
Going back to ancient discussions on a trim function approximately 15
years ago, the thought then was that Unicode whitespace characters would
need to be included in a default implementation.  Just including ASCII
whitespace wasn't good enough back then even.

**Comment by Grinnz** [comment]
I think as mentioned including the examples of the four most common ASCII whitespace characters is beneficial. I also believe the equivalent regexp operation is important to include as it is reference documentation and should endeavor to exactly specify what it will do (barring undefined behavior).

**Comment by Grinnz** [comment]
To summarize IRC discussion regarding the optional argument specifying what to trim:

1. pass a regex, this has the benefit of having an easy to understand specification (pass as `qr//`) but the downside of allowing crazy behavior we shouldn't support and almost impossible to optimize.
2. pass a list of characters, this is relatively simple and easy to optimize but doesn't allow you to pass something like `\s#` to trim whitespace characters (incl. unicode) and hash marks, you have to pass the specific characters you want to trim.
3. pass a character class, this would be the most reasonable option logically, but this has the problem that unlike a regex we don't have a straightforward firstclass way to specify character classes, so you would have to pass it as a string. it's annoying and weird to specify something like the `\s` character class in a string since you would have to double the backslash for it to be passed through until it's used as a regex.
4. don't allow specifying the optional argument. this avoids the whole discussion and relies on that the use case where someone wants to customize what characters to trim is small enough that those isolated instances can just use the equivalent regexes presented in docs and tweak them to their needs.

**Comment by leonerd** [comment]
@Grinnz Re point 3: you can always use `'\s'`

**Comment by scottchiefbaker** [comment]
My vote is 2... Keep is simple and only strip **specific** characters, with no fancy regexps allowed. A real world example I've used in PHP is to trim the `|` off of Markdown tables:

    $str = trim("|Hello|", "|");

Barring that I'd go with 4.

**Comment by Grinnz** [comment]
@leonerd but then you want to use the double quote for actual character escapes like `"\0"` or so and it seems like confusion waiting to happen.

**Comment by leonerd** [comment]
Honestly at this point, seeing all the options tempts me into option 4- just don't make it configurable. "It just uses [\s]" is easy to explain and document.

**Comment by scottchiefbaker** [comment]
Languages that allow a second argument list of what to remove: PHP

Languages that just implement `\s`: Javascript, Ruby, Raku, Python

**Comment by chrestomanci** [comment]
I think option 4 is best, as makes it easy for novice programmers to use and it allows us to optimise the common use case, and make sure it stays correct as the Unicode standards evolve, and they change what they mean by white space.

If someone has specific requirements to trim something else, then they are probably advanced enough to write their own regex to suit their needs, and they can deal with the corner cases.

**Comment by Leont** [comment]
I'm not so sure we should do this, for two reasons.

1. Why would we want this to be a keyword as opposed an ordinary function (in a module)? I can see several downsides to it (e.g. it wouldn't be available to older perls) and no necessity. If we add everything to core than could be useful, we run the risk of ending up with 3000 builtins like PHP did. We already have quite a lot (200+ AFAIK), we should be careful about adding more.

2. Our general practice whenever possible is to first implement on CPAN, and only if it's successful incorporate it in core. That step seems to be skipped here. Actually there are CPAN modules (e.g. `String::Util`), but they don't seem much used.

**Comment by scottchiefbaker** [comment]
[Text::Trim](https://metacpan.org/pod/Text::Trim) does indeed already exist, and has for quite a while. I believe `trim()` is such a **core** feature of any string processing language, that it's criminal that we don't have one in core. Literally every other major language (see above) has one in core. Perl needs `trim()`.

We need to make Perl more approachable to novices. To me that means including some syntactical sugar things in core, even if they can be done with a regexp or via a module.

**Comment by scottchiefbaker** [comment]
After sleeping on it I think it makes more sense to move forward with only stripping whitespace and not other characters. Along those lines here is my updated doc template:

## trim

`trim` EXPR  

Strip whitespace from the beginning and end of a string.

This function returns a string with whitespace stripped from the beginning and end of string. `trim()` will remove these characters: 

" ",  an ordinary space.  
"\t", a tab.  
"\n", a new line (line feed).  
"\r", a carriage return.  

#### Examples:

     $str = trim("  Hello world");   # Returns "Hello world"
     $str = trim("\tHello world\n"); # Returns "Hello world"

**Comment by Leont** [comment]
> After sleeping on it I think it makes more sense to move forward with only stripping whitespace and not other characters. 

Quite frankly, any successful feature in the past more than decade was simple enough that you could explain them in one sentence (e.g. `//`, `s///r`, `say`), whereas failed features tend to be complicated (e.g. smartmatch). I would very much encourage to keep it simple.

> trim() will remove these characters:

We've consistently used `\s` for this purpose for decades, making this operation ASCII based in 2020 would be outright silly.

**Comment by leonerd** [comment]
Could we solve this by pulling `String::Util` or `Text::Utils` or somesuch in as a dual-life core module?

I kinda agree with @Leont - there's no reason this has to be a core *keyword* as opposed to just some function loaded by a module. If it was a core-shipped dual-life module that would be fine.

**Comment by xenu** [comment]
>Why would we want this to be a keyword as opposed an ordinary function (in a module)? I can see several downsides to it (e.g. it wouldn't be available to older perls)

@Leont: "it won't be available on older perls" isn't a valid argument because it applies to literally all new features.

> If we add everything to core than could be useful, we run the risk of ending up with 3000 builtins like PHP did. We already have quite a lot (200+ AFAIK), we should be careful about adding more.

It's true we have lots of keywords that shouldn't have been made built-in (socket functions, `dbmopen` `getpwnam` etc.) but they are a relic from the pre-CPAN era. It doesn't mean perl core doesn't lack features. In fact, compared to the other dynamic languages, like Python or Ruby, Perl's core is surprisingly poor. It's impossible to do even the most basic operations without using libraries. Half of List::Util belongs in the core language.

IMO adding `trim` is a great idea. Personally I prefer the version that doesn't accept any additional arguments and always trims `\s`.

**Comment by scottchiefbaker** [comment]
I have created a branch with tests and documentation to serve as a starting point for @leonerd to begin his official implementation.

https://github.com/scottchiefbaker/perl5/commits/trim

**Comment by leonerd** [comment]
Latest update: I have put an initial implementation in that satisfies the tests, though does generate one `undef` warning.

Several other core tests are currently upset about the new keyword (e.g. the `B::Deparse` ones) but fixing those is just the usual routine for any new core stuff.

**Comment by druud** [comment]
It always amazes me when someone implements a trim() by treating the start of the string first, instead of doing the tail first. But I didn't benchmark anything.

In general I trim a value in-place, like:
`s/\s+\z//, s/\A\s+//, for $s;  # AllTrim`

**Comment by leonerd** [comment]
@druud Because of the OOK mechanism in SVs, there's no performance requirement to trim the start before the end. Equivalent operations will be performed in either case.

**Comment by learnbyexample** [comment]
> Languages that allow a second argument list of what to remove: PHP
> 
> Languages that just implement `\s`: Javascript, Ruby, Raku, Python

Python does allow specifying characters to be trimmed, default is whitespace. [Example from documentation](https://docs.python.org/3/library/stdtypes.html#str.strip):

```python
>>> 'www.example.com'.strip('cmowz.')
'example'
```

Also, would the proposed `trim()` function support feature similar to `rstrip` and `lstrip`?

**Comment by druud** [comment]
I strongly prefer to put such functionality into a namespace like `String::Core`. Which then can be loaded by a module `common.pm` or such.
```
  use common 'string';
```

Be aware that `trim` means something else in many cases and contexts. For example, there probably needs to be a `unify_whitespace_inplace` variant too, which returns the length-change in non-void context.
There can be `LTrim` and `RTrim`, that only remove any whitespace characters for either end. And then `AllTrim`. Also often desired is to normalize (or unify) any internal whitespace run to a single `SPC`.

A string could remember that it already has been trimmed, such that a repeated call can be a NOP, by implementing a `had_trim` flag. The flags would be (p)reset at each change of the string value.

I'd rather move out `chop()` and `chomp()`, than add `trim()`.

Also consider code like:
```
  while (<>) {
    my $line= s/\s+\z//r;  # x-chomp
    ...
```
which could use a substr-optimisation:
```
   my $line= substr($_, 0, /.*\S/ ? $+[0] : 0);
```

**Comment by Astara** [comment]
On 2020/11/23 16:51, Dan Book wrote:
> > On Mon, Nov 23, 2020 at 7:46 PM L A Walsh <Astara@tlinx.org 
> > <mailto:Astara@tlinx.org>> wrote:
> >
> >     On 2020/11/18 08:37, Dan Book wrote:
> >     > On Wed, Nov 18, 2020 at 3:17 AM L A Walsh <Astara@tlinx.org
> >     <mailto:Astara@tlinx.org>
> >     > <mailto:Astara@tlinx.org <mailto:Astara@tlinx.org>>> wrote:
> >     > The documentation added by the PR explains it. It removes leading
> >     and
> >     > trailing whitespace.
> >     ----
> >         That begs the questions:
> >         0) if whitespace, that includes tabs, LF's CR's, i.e.
> >            any thing '\s' matches?
> >         1) What if I only want to trim at beginning or end?
> >         2) Why not support a regex like, say, TCL does?
> >
> >         I.e. why should a perl command be limited to only trimming
> >     whitespace?
> >
> >
> > This has been discussed in the feature request ticket, summed up 
> > at https://github.com/Perl/perl5/issues/17952#issuecomment-658462652.
----
    I didn't see anyone pointing out the circular logic of limiting
this to '\s'?  I.e.


      * scottchiefbaker <https://github.com/scottchiefbaker> * commented
      Jul 14, 2020
      <https://github.com/Perl/perl5/issues/17952#issuecomment-658467533>

Languages that allow a second argument list of what to remove: PHP

Languages that just implement |\s|: Javascript, Ruby, Raku, Python


I know for certain that Javascript was heavily influenced by perl,
Raku was perl6, python's regex (thus '\s' ) came from perl, and I'm pretty
sure Ruby had similar influences. 

Might want to add TCL to languages that support alternatives to simple
whitespace stripping.  TAB is oft used as a field separator, and skipping
past LF's seems and other white space seems unnatural.

That the perl-derived langs all used '\s' just shows how they got lazy
and used '\s' instead of a subset or just literal spaces.

Certainly support for an optional regex 'range', would be of more use, i.e.
[\ \t\j\r...].  Option 4 could be special cased easily enough.


So what was the objection to this being an '.xs' module in CORE?  Wouldn't
the same efficiencies be achieved?

**Comment by wchristian** [comment]
@druud
> I strongly prefer to put such functionality into a namespace like `String::Core`. Which then can be loaded by a module `common.pm` or such.

No.

`trim` must be in Perl core, activated by a feature pragma.

The point is to minimize the work a non-CPAN developer has to do to have access to the function.

Precisely because it's implementation is so small, in the range of the code needed for `say`, it must also be able to be used with minimum effort.

Any suggestion that involves tracking dependencies beyond Perl version, would increase the work-to-use needed to a point where the designated users would just not use it and quickly write it on their own, in a myriad ways. (I've seen a codebase with 20 unique implementations of trim.)

`say` is also "just a function" and is extremely similar in many ways. To treat `trim` any different from `say` requires extremely strong arguments to explain how it is different from `say`.

--

As for the separate question of "options?!":

That spends a lot of time thinking and talking about how to handle edge cases that rarely occur. In fact, the more often occurring edge cases i see in people using trim haven't even been considered in the conversation at all yet. (And i will not introduce them.) So frankly, the specific edge cases of "trim non-`/s` chars is the most edge.

Any such implementation would come at a cost (clearly documented in earlier replies) to the main target audience, and arguably the correct answer is:

If you're doing something this fancy with your trim, it should be well thought-through and documented and as such for that specific case write-your-own or load-from-cpan is the better answer.

@leonerd or @scottchiefbaker might feel differently on that though.

**Comment by wchristian** [comment]
@druud  
> A string could remember that it already has been trimmed, such that a repeated call can be a NOP, by implementing a `had_trim` flag. The flags would be (p)reset at each change of the string value.

I have a feeling that there is no easy way to add a "reset this flag on change" hook. Can you demonstrate that one exists?

**Comment by richardleach** [comment]
> > A string could remember that it already has been trimmed, such that a repeated call can be a NOP, by implementing a `had_trim` flag. The flags would be (p)reset at each change of the string value.
> 
> I have a feeling that there is no easy way to add a "reset this flag on change" hook.

That's my gut feeling too. Assuming that's the case, then even if perl's own ops were modified (adding a bit more complexity to each one), non-core XS modules wouldn't respect the flag without modification.

Other implementation thoughts:
* Where would this flag live? AFAIK there aren't a load of SvFLAGs going spare.
* When non-COW strings are copied, should the flag be propagated?
* Presumably the flag's value would be lost when, for example, a string is round-tripped as a hash key.
* Not sure the determination that a repeated call should be a NOP could be done at compile time, so there'd still be a runtime cost of OP dispatch + check for magic + check for the flag.

Also, if the spec of _trim_ ever changed (e.g. went from a pre-canned set of characters to allowing user-specified characters), or there became other types of trim (e.g. _ltrim_, _rtrim_), the flag's meaning would become more complicated.

**Comment by druud** [comment]
> @druud
> > I strongly prefer to put such functionality into a namespace like `String::Core`. Which then can be loaded by a module `common.pm` or such.
> 
> No.

I still prefer strawberry.

> `trim` must be in Perl core, activated by a feature pragma.

ITYM: Both `chop` and `chomp` must be removed from the Perl code, activated by a feature pragma.

> The point is to minimize the work a non-CPAN developer has to do to have access to the function.
```
  use common 'string';
```

> Precisely because it's implementation is so small, in the range of the code needed for `say`, it must also be able to be used with minimum effort.

There are many examples of commonly applied functionality with a small implementation.
For floor() you can use int(), but for ceil() you need to use POSIX or similar. For banker's rounding, you need to go elsewhere, like `0+ sprintf "%.0f", $v`. For limiting the length of a string, with some kind of `...` appended, somewhere else again. Etc. Etc.

> Any suggestion that involves tracking dependencies beyond Perl version, would increase the work-to-use needed to a point where the designated users would just not use it and quickly write it on their own, in a myriad ways. (I've seen a codebase with 20 unique implementations of trim.)
```
  use common 'io', 'math', 'string','types';
```

> `say` is also "just a function" and is extremely similar in many ways. To treat `trim` any different from `say` requires extremely strong arguments to explain how it is different from `say`.

I now wonder how equivalent those 20 implementations of `trim` where.

Checking a local code base:
```
$ git grep -E '^\s*sub\s+trim\b' lib |wc -l
16
```
of which about 12 do: `s/^\s+|\s+$//g`
which is likely inspired by `perldoc -q space`.


All in all, I still prefer to have something like a Common::String module, that you can load (in any recent version of Perl) with
```
  use common 'string';
```
which by default exports base-routines like `trim`, and can export more on demand, like `trim_inplace` and `unify_ws`.
```
  use common 'string=:ws,:utf8';
```

In Common::String, I prefer a lean `trim` that only removes `\s+` from start and end.
```
  use re '/aa';
  use common 'string';
  ...
  my @trimmed= trim( @raw );
```

And `trim` should probably not touch the tainted status.

-- Ruud

**Comment by wchristian** [comment]
> @druud
> ITYM: Both `chop` and `chomp` must be removed from the Perl code, activated by a feature pragma.

This demonstrates the main issue with your comments: You fail to even grasp what i wrote and counter with snark and sarcasm, which has absolutely no place in a conversation such as this. The correct way to go about this is: If you in any way feel that sarcasm is called for, ask to ensure that your understanding is accurate instead.

And to demonstrate directly to you why the comment above is in "not even wrong" territory: feature pragmata are used for features changes that change perl parsing. Your request above would only make sense as "add a feature pragma that removes chop and chomp".

> There are many examples of commonly applied functionality with a small implementation.

The examples you gave are used less often, however, feel free to provide similar patches as with this one. The goal is to make perl easier to use for basic tasks.

> All in all, I still prefer to have something like a Common::String module, that you can load (in any recent version of Perl) with

Please enjoy: https://metacpan.org/pod/String::Util

--

@richardleach Excellent demonstration of the overwhelming complexity that blocks that particular idea.

**Comment by druud** [comment]
> > @druud
> > ITYM: Both `chop` and `chomp` must be removed from the Perl code, activated by a feature pragma.
> 
> This demonstrates the main issue with your comments: You fail to even grasp what i wrote and counter with snark and sarcasm, which has absolutely no place in a conversation such as this. The correct way to go about this is: If you in any way feel that sarcasm is called for, ask to ensure that your understanding is accurate instead.

I just copied your style, that started with a 'No.' So please go and find a mirror to look into now.

> > All in all, I still prefer to have something like a Common::String module, that you can load (in any recent version of Perl) with
> 
> Please enjoy: https://metacpan.org/pod/String::Util

I had a quick look, but its implementation of trim() is not so good.
Text::Trim put more thought into it; IMO a bit too much.

-- Ruud

**Comment by wchristian** [comment]
To be unmistakable: Use neither snark nor sarcasm. None was present before, and it has no place here.

If you have serious factual objections to the PR as given, please give them, or questions about it, please ask them.

Otherwise there currently is no reason not to proceed, as it perfectly fulfills the necessities of the situation while providing a clear advantage in the same vein as `say` did.

**Comment by scottchiefbaker** [comment]
I have update my branch, and thus the PR, with a fully merge-able and fully test compliant version of `trim()` in core.

**Comment by salva** [comment]
@scottchiefbaker ,

> Languages that just implement `\s`: Javascript, Ruby, Raku, Python

Python accepts as an argument the list of characters to remove and it also provides `rstrip` and `lstrip` variants.

https://docs.python.org/3/library/stdtypes.html?highlight=strip#str.strip

**Comment by rwp0** [comment]
Very sad to see the pull request and the issue to be closed 🙁.

**Comment by esabol** [comment]
> Very sad to see the pull request and the issue to be closed 🙁.

What happened here?

EDIT: Oh, I see it was merged in PR #19433.
